internal/terraform/transform_query.go

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package terraform
+
+import (
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/dag"
+)
+
+type QueryTransformer struct {
+	// queryPlan is true when we are planning list resources.
+	queryPlan bool
+
+	// if validate is true, we are in validate mode and should not exclude any resources.
+	validate bool
+}
+
+func (t *QueryTransformer) Transform(g *Graph) error {
+	if !t.queryPlan {
+		// if we are not running a query-specific operation, we don't need to transform the graph
+		// as query-related files would not have been part of the parsed config.
+		return nil
+	}
+
+	if t.validate {
+		// if we are validating query files, we validate all resources
+		return nil
+	}
+
+	for v := range dag.SelectSeq[GraphNodeConfigResource](g.VerticesSeq()) {
+		// we only get here if we are building a query plan, but not validating.
+		// Because the config would contain resource blocks from traditional .tf files,
+		// we need to exclude them from the plan graph.
+		// If the node is to be removed, we need to remove it and its descendants from the graph.
+		if v.ResourceAddr().Resource.Mode != addrs.ListResourceMode {
+			deps := g.Descendants(v)
+			g.Remove(v)
```

Comment:
I think you want to reverse the logic here and find what you need to keep rather than remove -- you can see the current `TargetsTransformer` for an example. While it's unlikely to happen between resources right now with the structure of query files, this can technically remove things like named values or providers you depend on, because those can still reference non-list resources.

Reply:
The correct way to evaluate dependencies is to follow the dependencies from the source (usually via `Ancestors`). If you walk in the opposite direction, you don't know where those nodes in turn have their own dependencies. This means that a nodes which is a shared dependency of what you want to keep and what you want to remove could be removed. Take this minimal example

```
ephemeral test_secret key {
}

provider "bufo" {
  key = ephemeral.test_secret.key
}

list bufo_resource all {
}
```

When you reach `ephemeral.test_secret_key` you would back backwards and remove the bufo provider which the list resource depends on, and actually remove the list resource too! What instead needs to happen is that you start from the list resource and keep everything that it depends on.

Reply:
Yeah, technically the provider can reference managed resources too, although of course it's not recommended. If the query is being executed when that resource already exists in the state, then will probably still work; but if the value is unknown during the plan, then there's not much we can do other than error out.